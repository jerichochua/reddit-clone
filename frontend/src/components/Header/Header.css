.header {
  background-color: #cee3f8;
  height: 48px;
  display: flex;
  padding: 0 10vw;
  margin-bottom: 24px;
  border-bottom: 1px solid #5f99cf;
  z-index: 99;
}

.header-logo {
  font-weight: 500;
  display: flex;
  align-items: center;
  font-size: 24px;
  padding: 0 16px;
  margin-right: auto;
  text-decoration: none;
  color: black;
}

.header-item {
  display: flex;
  align-items: center;
  padding: 0 16px;
}

.header-link {
  cursor: pointer;
  text-decoration: none;
  color: black;
}

.dark-mode-icon {
  width: 20px;
  height: 20px;
}

.dark-mode-btn {
  display: flex;
  align-items: center;
  padding: 0 8px;
  cursor: pointer;
}

@media (max-width: 1024px) {
  .header {
    padding: 0 5vw;
  }
}

@media (max-width: 768px) {
  .header {
    padding: 0;
  }
}

@media (max-width: 425px) {
  .header {
    margin-bottom: 16px;
    height: 40px;
  }

  .header-logo {
    padding: 0 8px 0 16px;
    font-size: 18px;
  }

  .header-item {
    padding: 0 8px;
    font-size: 14px;
  }
}
