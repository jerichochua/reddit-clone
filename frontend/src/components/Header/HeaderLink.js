import React from 'react';

const HeaderLink = ({ text }) => <p className='header-link'>{text}</p>;

export default HeaderLink;
