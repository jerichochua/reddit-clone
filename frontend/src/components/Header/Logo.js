import React from 'react';
import { Link } from 'react-router-dom';

const Logo = () => (
  <Link className='header-logo' to='/'>
    not-reddit
  </Link>
);

export default Logo;
