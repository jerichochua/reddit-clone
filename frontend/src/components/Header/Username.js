import React from 'react';

const Username = ({ username }) => {
  return <div className='header-item'>logged in as {username}</div>;
};

export default Username;
