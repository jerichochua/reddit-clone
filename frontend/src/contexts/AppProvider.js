import React, { createContext, useContext, useReducer } from 'react';
import jwt_decode from 'jwt-decode';

const AppContext = createContext();
AppContext.displayName = 'AppContext';

const token = localStorage.getItem('token');
const userId = token ? jwt_decode(token).id : null;
const username = token ? jwt_decode(token).username : null;

const initialState = {
  userId,
  username,
  token,
};

const reducer = (state, action) => {
  switch (action.type) {
    case 'SET_USER':
      return {
        ...state,
        userId: action.payload.userId,
        username: action.payload.username,
      };
    case 'SET_TOKEN':
      return { ...state, token: action.payload };
    case 'LOGOUT':
      return { ...state, userId: null, username: null, token: null };
    default:
      return state;
  }
};

export const AppProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  return (
    <AppContext.Provider value={{ state, dispatch }}>
      {children}
    </AppContext.Provider>
  );
};

export const useAppContext = () => {
  return useContext(AppContext);
};
