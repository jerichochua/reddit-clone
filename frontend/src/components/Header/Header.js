import React from 'react';
import HeaderLink from './HeaderLink';
import Logo from './Logo';
import Username from './Username';
import { useAppContext } from '../../contexts/AppProvider';
import './Header.css';

const Header = () => {
  const { state, dispatch } = useAppContext();
  const user = state.user;

  const logout = () => {
    localStorage.removeItem('token');
    dispatch({ type: 'LOGOUT' });
  };

  return (
    <div className='header'>
      <Logo />
      {user ? (
        <>
          <Username username={user} />
          <HeaderLink text='create post' to='createpost' />
          <span className='header-item header-link' onClick={logout}>
            logout
          </span>
        </>
      ) : (
        <>
          <HeaderLink text='log in' to='login' />
          <HeaderLink text='sign up' to='signup' />
        </>
      )}
    </div>
  );
};

export default Header;
