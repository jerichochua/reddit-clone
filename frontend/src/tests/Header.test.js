import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { AppProvider } from '../contexts/AppProvider';
import Header from '../components/Header/Header';

const initialState = {
  userId: '1',
  username: 'user',
  token: 'token',
};

it('renders correctly', () => {
  const component = renderer.create(
    <MemoryRouter>
      <AppProvider>
        <Header />
      </AppProvider>
    </MemoryRouter>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders logo text', () => {
  const component = renderer.create(
    <MemoryRouter>
      <AppProvider>
        <Header />
      </AppProvider>
    </MemoryRouter>
  );
  const tree = component.toJSON();
  expect(tree.children[0].children[0]).toBe('not-reddit');
});

it('renders log in and sign up links when not logged in', () => {
  const component = renderer.create(
    <MemoryRouter>
      <AppProvider>
        <Header />
      </AppProvider>
    </MemoryRouter>
  );
  const tree = component.toJSON();
  expect(tree.children[1].children[0]).toBe('log in');
  expect(tree.children[2].children[0]).toBe('sign up');
});

it('renders username, create post, and logout links when logged in', () => {
  const component = renderer.create(
    <MemoryRouter>
      <AppProvider value={initialState}>
        <Header />
      </AppProvider>
    </MemoryRouter>
  );
  const tree = component.toJSON();
  expect(tree.children[1].children[0]).toBe('user');
  expect(tree.children[2].children[0]).toBe('create post');
  expect(tree.children[3].children[0]).toBe('logout');
});
