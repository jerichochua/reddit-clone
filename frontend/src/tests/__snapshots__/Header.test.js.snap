// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="header"
>
  <a
    className="header-logo"
    href="/"
    onClick={[Function]}
  >
    not-reddit
  </a>
  <a
    className="header-item header-link"
    href="/login"
    onClick={[Function]}
  >
    log in
  </a>
  <a
    className="header-item header-link"
    href="/signup"
    onClick={[Function]}
  >
    sign up
  </a>
</div>
`;
